pkg/tbtc/tbtc.go

```diff
@@ -125,9 +136,36 @@ func Initialize(
 		}()
```

Comment:
There is one more thing that occurred to me: at line 112 in this file (not modified in this PR) we have:
```
_ = chain.OnDKGStarted(func(event *DKGStartedEvent) {
  go func() {
    if ok := deduplicator.notifyDKGStarted(
      event.Seed,
    ); !ok {

    // (...)
    
    node.joinDKGIfEligible(
      event.Seed,
      event.BlockNumber,
    )
  }()
})
```

I am pretty sure this will be problematic on mainnet as-is because of the fact `event.BlockNumber` is a part of the signature:

https://github.com/keep-network/keep-core/blob/b8dcba4a37e68febff082bd2c6282b5fac1f7d04/solidity/ecdsa/contracts/libraries/EcdsaDkg.sol#L103-L106

We will have small chain reorganizations on mainnet resulting in `DKGStartedEvent`s emitted one after another, with different `startBlock`. They will be deduplicated by `deduplicator.notifyDKGStarted` but the `startBlock` matters later for the on-chain result submission.

What I would suggest doing here is to wait for 20 blocks (this feels like safe enough finalization for Ethereum) then take the last, `GetPastDkgStartedEvent`, and trigger DKG based on the values from that event.

Reply:
Yeah, good point. Just a couple of questions:
- Are we sure `GetPastDkgStartedEvent` will return the same event for all? I can imagine a case when not all members receive the same due to ETH node re-connections or slight state differences
- What about deduplication? Do we want to keep the current logic? Do we want to compare the original event with the one fetched via `GetPastDkgStartedEvent`?

Reply:
> Are we sure GetPastDkgStartedEvent will return the same event for all? I can imagine a case when not all members receive the same due to ETH node re-connections or slight state differences

If we wait for 20 blocks then I am pretty sure the answer is yes. This function, under the hood, is just fetching filtered logs from the contract. It's a simple call to Ethereum node, it does not maintain a state. If the smart contract's state has enough confirmations than all nodes should receive the same result.

> What about deduplication? Do we want to keep the current logic? Do we want to compare the original event with the one fetched via GetPastDkgStartedEvent?

I think we need to deduplicate to not start multiple monitoring loops for 20 confirmations. I don't think we need to compare events, I think we should just accept the fact the event can change and use the one fetched after the confirmations. 

In my mind, this works as: 
- I know DKG is probably going to happen. So I start just one (!) monitoring.
- I know the information I have right now may change, I need to wait 20 blocks to see the final version.
- After 20 blocks, I confirm if DKG is in progress by checking the event. I take that event as a final source of truth.

Reply:
Ok, I'll rework that.

Reply:
I spent some time on it and I'm afraid this is not so simple. If we wait for 20 confirmation blocks, we start the actual off-chain protocol with a delay that breaks a lot of things. This is because we use the original block number from the event as the starting point but the actual block is more than 20 blocks in the future. This PR is already a ball of mud so I think we should think about it and address that problem separately. I created an issue capturing it https://github.com/keep-network/keep-core/issues/3456